CB-7212: Add datalake service changes to perform database backup/restore.

kkalvagadda1 opened the pull request:
Changes needed to the datalake service to handle the requests to perform backup and restore of the database backing datalake. This includes exposing API's to perform/restore and also API's to poll the status of those operations. These API internally uses separate flows for backup and restore operations. 
Backup and restore operation return UUID which can be used to track the status of the operation requested.

I have added tests for the service backing the backup/restore operations.

Performed manual testing to make sure the Flow transition is done properly. Here are the state transitions are listed in the sheet below
https://docs.google.com/spreadsheets/d/1rms3b-YGAIeTLf-Pqm4gSnJopoQGQ1O8pyTVRR_86t0/edit?usp=sharing

brycederriso commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseDrState.java:
```diff
@@ -0,0 +1,33 @@
+public enum DatalakeDatabaseDrState implements FlowState {
+
+    INIT_STATE,
+    DATALAKE_DATABSE_BACKUP_START_STATE,
```
**nitpick:** This should be `DATABASE`.

hreeve-cloudera commented on core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/common/Status.java:
```diff
@@ -12,7 +12,13 @@
     AVAILABLE(StatusKind.FINAL),
     UPDATE_IN_PROGRESS(StatusKind.PROGRESS),
     UPDATE_REQUESTED(StatusKind.PROGRESS),
+    BACKUP_REQUESTED(StatusKind.PROGRESS),
```
This is a general comment for the review about how the status works on the cloudbreak side. It might or might not have any affect on what you do with status in your code. But on the cloudbreak side, there is no separate states of requested/started and in-progress. By the time the flow advances to the point where it's updating the stack status, the operation is already in progress. So in my version of these status messages, I had:
`
    BACKUP_IN_PROGRESS(StatusKind.PROGRESS),
    BACKUP_FINISHED(StatusKind.FINAL),
    BACKUP_FAILED(StatusKind.FINAL),
    RESTORE_IN_PROGRESS(StatusKind.PROGRESS),
    RESTORE_FINISHED(StatusKind.FINAL),
    RESTORE_FAILED(StatusKind.FINAL);
`
I see you have multiple places where you have a separate start state, and an in-progress state. For the datalake service I think that makes sense, since you can mark it as started as soon as you get the request, and in-progress when you get the flow identifier back from cloudbreak. I just wanted to give you a heads up that cloudbreak won't be using the REQUESTED stattus.

kkalvagadda1 replied:
I just followed the states defined for UPDATE. UPDATE has UPDATE_REQUESTED and UPDATE_IN_PROGRESS etc. 

I can update the patch accordingly.

hreeve-cloudera commented on core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/StackV4Endpoint.java:
```diff
@@ -269,4 +273,17 @@ FlowIdentifier setClusterMaintenanceMode(@PathParam("workspaceId") Long workspac
     @ApiOperation(value = UPDATE_SALT, nickname = "updateSaltByName")
     FlowIdentifier updateSaltByName(@PathParam("workspaceId") Long workspaceId, @PathParam("name") String name);
 
+    @POST
```
To match REST conventions and StackV4Endpoint naming conventions:
```suggestion
    @POST
    @Path("{name}/database_backup")
    @Produces(MediaType.APPLICATION_JSON)
    @ApiOperation(value = DATABASE_BACKUP, nickname = "databaseBackup")
    BackupV4Response backupDatabaseByName(@PathParam("workspaceId") Long workspaceId, @PathParam("name") String name,
            @QueryParam("location") String location, @QueryParam("backupId") String backupId);

    @POST
    @Path("{name}/database_restore")
    @Produces(MediaType.APPLICATION_JSON)
    @ApiOperation(value = DATABASE_RESTORE, nickname = "databaseRestore")
    RestoreV4Response restoreDatabaseByName(@PathParam("workspaceId") Long workspaceId, @PathParam("name") String name,
            @QueryParam("location") String location, @QueryParam("backupId") String backupId);
```

kkalvagadda1 replied:
Basically the suggestion is to accepted location and backup-id as query parameters, right.

pdarvasi commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    private AttemptResult<StackV4Response> sdxDrSucceeded(StackV4Response stackV4Response) {
+        LOGGER.info("Database DR operation doe SDX cluster {} is successfull", stackV4Response.getCluster().getName());
```
typo: successful + doe -> for

kkalvagadda1 replied:
will do

pdarvasi commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+     */
+    public void updateDatabaseStatusEntry(String operationId, SdxDatabaseDrStatus.Status status, String failedReason) {
+        if (Strings.isNullOrEmpty(operationId)) {
+            return;
```
we might log this as it should not happen

kkalvagadda1 replied:
will do.

pdarvasi commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+     * @return Backup status
+     */
+    public SdxDatabaseBackupStatusResponse getDatabaseBackupStatus(SdxCluster sdxCluster, String operationId) {
+        SdxDatabaseDrStatus drStatus = sdxDatabaseDrStatusRepository.findSdxDatabaseDrStatusByOperationId(operationId);
```
the first 6 lines of getDatabaseBackupStatus and getDatabaseRestoreStatus are almost the same -> should be extracted to a common method with SdxDatabaseDrStatusTypeEnum as parameter

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/SdxDatabaseDrStatus.java:
```diff
@@ -0,0 +1,102 @@
+import javax.validation.constraints.NotNull;
+
+@Entity
+@Table(name = "sdxDatabaseDrstatus", uniqueConstraints = @UniqueConstraint(columnNames = {"operationId"}))
```
you shouldn't specify the name here

kkalvagadda1 replied:
This patch also has the sql file to create table "sdxDatabaseDrstatus". Without having the name here, how can I assume that the entity would using the table with the name "sdxDatabaseDrstatus"?

kkalvagadda1 replied:
will do.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/SdxDatabaseDrStatus.java:
```diff
@@ -0,0 +1,102 @@
+@Entity
+@Table(name = "sdxDatabaseDrstatus", uniqueConstraints = @UniqueConstraint(columnNames = {"operationId"}))
+public class SdxDatabaseDrStatus {
+    public enum SdxDatabaseDrStatusTypeEnum {
+        BACKUP,
+        RESTORE
+    }
+
+    public enum Status {
+        INIT,
+        TRIGGERRED,
+        INPROGRESS,
+        SUCCEEDED,
+        FAILED
+    }
```
please move them out from the entity

kkalvagadda1 replied:
will do.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/SdxDatabaseDrStatus.java:
```diff
@@ -0,0 +1,102 @@
+    private Long id;
+
+    @NotNull
+    @Enumerated(EnumType.STRING)
```
please use converter

kkalvagadda1 replied:
Can you elaborate?

kkalvagadda1 replied:
Figured it out. will do.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/SdxDatabaseDrStatus.java:
```diff
@@ -0,0 +1,102 @@
+    private String statusReason;
+
+    @NotNull
+    @Enumerated(EnumType.STRING)
```
here too

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseBackupActions.java:
```diff
@@ -0,0 +1,173 @@
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, DatalakeDatabaseBackupStartEvent payload, Map<Object, Object> variables) throws Exception {
```
`throws Exception` is not necessary. same for all the `doExecute`

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/event/DatalakeDatabaseBackupCouldNotStartEvent.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String selector() {
+        return "DatalakeDatabaseBackupCouldNotStartEvent";
```
I won't go through all of your request, but please check them all.
The parent class already has defined this with `EventSelectorUtil` so you should depend on that instead of overriding it with the same

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseDrEvent.java:
```diff
@@ -0,0 +1,32 @@
+
+public enum DatalakeDatabaseDrEvent implements FlowEvent {
+
+    DATALAKE_DATABASE_BACKUP_EVENT("DATALAKE_DATABASE_BACKUP_EVENT"),
+    DATALAKE_DATABASE_BACKUP_COULD_NOT_START_EVENT("DATALAKE_DATABASE_BACKUP_COULD_NOT_START_EVENT"),
+    DATALAKE_DATABASE_BACKUP_IN_PROGRESS_EVENT("DATALAKE_DATABASE_BACKUP_IN_PROGRESS_EVENT"),
+    DATALAKE_DATABASE_BACKUP_SUCCESS_EVENT("DATALAKE_DATABASE_BACKUP_SUCCESS_EVENT"),
+    DATALAKE_DATABASE_BACKUP_FAILED_EVENT("DATALAKE_DATABASE_BACKUP_FAILED_EVENT"),
+    DATALAKE_DATABASE_BACKUP_FINALIZED_EVENT("DATALAKE_DATABASE_BACKUP_FINALIZED_EVENT"),
+    DATALAKE_DATABASE_BACKUP_FAILURE_HANDLED_EVENT("DATALAKE_DATABASE_BACKUP_FAILURE_HANDLED_EVENT"),
+    DATALAKE_DATABASE_RESTORE_EVENT("DATALAKE_DATABASE_RESTORE_EVENT"),
+    DATALAKE_DATABASE_RESTORE_COULD_NOT_START_EVENT("DATALAKE_DATABASE_RESTORE_COULD_NOT_START_EVENT"),
+    DATALAKE_DATABASE_RESTORE_IN_PROGRESS_EVENT("DATALAKE_DATABASE_RESTORE_IN_PROGRESS_EVENT"),
+    DATALAKE_DATABASE_RESTORE_SUCCESS_EVENT("DATALAKE_DATABASE_RESTORE_SUCCESS_EVENT"),
+    DATALAKE_DATABASE_RESTORE_FAILED_EVENT("DATALAKE_DATABASE_RESTORE_FAILED_EVENT"),
+    DATALAKE_DATABASE_RESTORE_FINALIZED_EVENT("DATALAKE_DATABASE_RESTORE_FINALIZED_EVENT"),
+    DATALAKE_DATABASE_RESTORE_FAILURE_HANDLED_EVENT("DATALAKE_DATABASE_RESTORE_FAILURE_HANDLED_EVENT");
```
please check my comment on one of your event, and use there and here also `EventSelectorUtil` instead of duplicating string everywhere

kkalvagadda1 replied:
I do not have classes defined for all the events. For the events for which classes are defined, i will make changes following your suggestion.

lacikaaa replied:
you don't have to create classes for all events. I didn't want to go through all of them to tag which should be replaced. Leaving those untouched which don't have class is fine

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/event/DatalakeDatabaseRestoreCouldNotStartEvent.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String toString() {
+        final StringBuffer sb = new StringBuffer("DatalakeDatabaseRestoreCouldNotStartEvent{");
```
you should use either plain String with concatenation or `StringBuilder`.
also final is unnecessary

kkalvagadda1 replied:
I will remove the final.
StringBuilder is not thread-safe by StringBuffer is. That is why I used StringBuffer.
Is there a specific reason to use StringBuilder?

lacikaaa replied:
why would you need thread-safe solution here? it has a performance drawback
also simple string concatenation would do, I usually use that for toString

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+ * Service to perform backup/restore of the database backing SDX.
+ */
+@Component
+public class SdxDrService {
```
this class needs some identation fix, idea fixed 8 lines for me

lacikaaa replied:
I think this class should be broken into 3
- backup
- restore
- common part
what do you think?

kkalvagadda1 replied:
Fixed the indentation. What is the rationale to split the service into multiple parts?

lacikaaa replied:
ideally one class does one thing, see: single-responsibility principle
also it would be easier to read as there would be less code in that class. so if you are looking for backup related stuff, you don't have to go through the restore part

kkalvagadda1 replied:
Will create a sperate patch for it.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+/**
+ * Service to perform backup/restore of the database backing SDX.
+ */
+@Component
```
if service is the name I think the annotation should be service

kkalvagadda1 replied:
will update

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+import com.sequenceiq.sdx.api.model.SdxDatabaseRestoreStatusResponse;
+
+/**
+ * Service to perform backup/restore of the database backing SDX.
```
if we need this comment then the name of the class is wrong

kkalvagadda1 replied:
Will rename the class to SdxDatabaseDrService.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+    private StackV4Endpoint stackV4Endpoint;
+
+    @Inject
+    private SdxStatusService sdxStatusService;
```
unused

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+    private CloudbreakFlowService cloudbreakFlowService;
+
+    @Inject
+    private DatabaseServerV4Endpoint databaseServerV4Endpoint;
```
unused

kkalvagadda1 replied:
will update.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+     */
+    public SdxDatabaseBackupStatusResponse getDatabaseBackupStatus(SdxCluster sdxCluster, String operationId) {
+        SdxDatabaseDrStatus drStatus = sdxDatabaseDrStatusRepository.findSdxDatabaseDrStatusByOperationId(operationId);
+        if ((drStatus == null) || (!drStatus.getSdxClusterId().equals(sdxCluster.getId()))
+                || (!drStatus.getOperationType().equals(SdxDatabaseDrStatus.SdxDatabaseDrStatusTypeEnum.BACKUP))) {
```
this condition is a bit hard to read and also almost the same as in `getDatabaseRestoreStatus`
it would worth to refactor them into a method

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+        StackV4Response stackV4Response = stackV4Endpoint.get(0L, sdxCluster.getClusterName(), Collections.emptySet());
+        LOGGER.info("Response from cloudbreak: {}", JsonUtil.writeValueAsString(stackV4Response));
+        ClusterV4Response cluster = stackV4Response.getCluster();
+        if (stackAndClusterAvailable(stackV4Response, cluster)) {
+            return sdxDrSucceeded(stackV4Response);
+        } else {
+            if (Status.BACKUP_FAILED.equals(stackV4Response.getStatus()) ||
+                    Status.RESTORE_FAILED.equals(stackV4Response.getStatus())) {
+                LOGGER.info("{} failed for Stack {} with status {}", pollingMessage, stackV4Response.getName(), stackV4Response.getStatus());
+                return sdxDrFailed(sdxCluster, stackV4Response.getStatusReason(), pollingMessage);
+            } else if (Status.BACKUP_FAILED.equals(stackV4Response.getCluster().getStatus()) ||
+                    Status.RESTORE_FAILED.equals(stackV4Response.getCluster().getStatus())) {
+                LOGGER.info("{} failed for Cluster {} status {}", pollingMessage, stackV4Response.getCluster().getName(),
+                        stackV4Response.getCluster().getStatus());
+                return sdxDrFailed(sdxCluster, stackV4Response.getCluster().getStatusReason(), pollingMessage);
+            } else {
+                if (FINISHED.equals(flowState)) {
+                    LOGGER.info("Flow finished, but Backup/Restore is not complete: {}", sdxCluster.getClusterName());
+                    return sdxDrFailed(sdxCluster, "stack is in improper state", pollingMessage);
+                } else {
+                    return AttemptResults.justContinue();
+                }
+            }
+        }
```
this if-else thing got out of hand. please refactor parts of into separate method with meaningful names so it would be easier to follow what happens

kkalvagadda1 replied:
will refactor it.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+                        String.format("SDX cluster with Id %d not found", clusterId));
+                throw notFound("SDX cluster", clusterId).get();
+            });
+        } catch (WebApplicationException e) {
+            String message = String.format("Database restore failed for datalake-id: [%d]. Message: [%s]", clusterId, e.getMessage());
+            throw new CloudbreakApiException(message);
+        }
```
you might want to use `WebApplicationExceptionMessageExtractor` here

kkalvagadda1 replied:
Will use it

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxController.java:
```diff
@@ -246,24 +250,31 @@ public FlowIdentifier stopByCrn(@ResourceCrn String crn) {
 
     @Override
     @CheckPermissionByAccount(action = AuthorizationResourceAction.BACKUP_DATALAKE)
     public SdxDatabaseBackupStatusResponse backupDatabaseStatusByName(@ResourceName String name, String operationId) {
```
will fix.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxController.java:
```diff
@@ -246,24 +250,31 @@ public FlowIdentifier stopByCrn(@ResourceCrn String crn) {
 
     @Override
     @CheckPermissionByAccount(action = AuthorizationResourceAction.RESTORE_DATALAKE)
     public SdxDatabaseRestoreStatusResponse restoreDatabaseStatusByName(@ResourceName String name, String operationId) {
```
will fix.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:
```diff
@@ -104,6 +109,18 @@ public FlowIdentifier triggerSdxStopFlow(SdxCluster cluster) {
         return notify(selector, new SdxStartStopEvent(selector, cluster.getId(), userId));
     }
 
+    public String triggerDatalakeDatabaseBackupFlow(Long sdxId, String backupId, String backupLocation) {
```
Why should FlowIdentifier return when it is not used? API invoking triggerDatalakeDatabaseBackupFlow uses operationId instead of flowIdentifier.

I can make changes to avoid the need for notifyDatabaseDrEvent. Does that address your concern?

kkalvagadda1 replied:
I have removed method "notifyDatabaseDrEvent" and re-used "notify".

kkalvagadda1 replied:
will update.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseBackupActions.java:
```diff
@@ -0,0 +1,173 @@
+
+            @Override
+            protected void doExecute(SdxContext context, SdxEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Datalake database backup is in progress for {} ", payload.getResourceId());
```
yes, you are right. I will fix it.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseBackupActions.java:
```diff
@@ -0,0 +1,173 @@
+                LOGGER.error("Datalake database backup failed for datalake with id: {}", payload.getResourceId(), exception);
+                String operationId = (String) variables.get(OPERATION_ID);
+                sdxDrService.updateDatabaseStatusEntry(operationId, SdxDatabaseDrStatus.Status.FAILED, exception.getLocalizedMessage());
+                sendEvent(context, DATALAKE_DATABASE_BACKUP_FINALIZED_EVENT.event(), payload);
```
Just for understanding sake, Is "DATALAKE_DATABASE_BACKUP_FINALIZED_EVENT" used only in success cases?

pdarvasi replied:
Yes, we use it like that

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseRestoreActions.java:
```diff
@@ -0,0 +1,170 @@
+            protected void doExecute(SdxContext context, SdxEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Datalake database restore is in progress for {} ", payload.getResourceId());
+                String operationId = (String) variables.get(OPERATION_ID);
+                sendEvent(context, DatalakeDatabaseRestoreWaitRequest.from(context, operationId));
```
yes, you are right. I will fix it.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseRestoreActions.java:
```diff
@@ -0,0 +1,170 @@
+                Exception exception = payload.getException();
+                LOGGER.error("Datalake database restore could not be started for datalake with id: {}", payload.getResourceId(), exception);
+                String operationId = (String) variables.get(OPERATION_ID);
+                sdxDrService.updateDatabaseStatusEntry(operationId, SdxDatabaseDrStatus.Status.FAILED, exception.getLocalizedMessage());
```
database backup/restore is one part of the datalake backup restore. There will be a separate patch to update the datalake status using SdxStatusService which will be invoked by datalakedr service which orchestrates the datalake backup and restore.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+
+    public void databaseRestore(SdxDatabaseDrStatus drStatus, Long clusterId, String backupId, String backupLocation) {
+        try {
+            sdxDatabaseDrStatusRepository.save(drStatus);
```
database backup/restore is one part of the datalake backup restore. There will be a separate patch to update the datalake status using SdxStatusService which will be invoked by datalakedr service which orchestrates the datalake backup and restore.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+                cloudbreakFlowService.saveLastCloudbreakFlowChainId(sdxCluster, restoreV4Response.getFlowIdentifier());
+                updateDatabaseStatusEntry(drStatus.getOperationId(), SdxDatabaseDrStatus.Status.TRIGGERRED, null);
+            }, () -> {
+                updateDatabaseStatusEntry(drStatus.getOperationId(), SdxDatabaseDrStatus.Status.FAILED,
```
will do.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:
```diff
@@ -132,4 +149,30 @@ private FlowIdentifier notify(String selector, SdxEvent acceptable) {
 
     }
+
+    String notifyDatabaseDrEvent(String selector,  DatalakeDatabaseDrStartBaseEvent acceptable) {
```
will remove the method.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseBackupActions.java:
```diff
@@ -0,0 +1,173 @@
+
+            @Override
+            protected void doExecute(SdxContext context, DatalakeDatabaseBackupFailedEvent payload, Map<Object, Object> variables) throws Exception {
+                Exception exception = payload.getException();
```
I understand that the code is duplicate. In fact I tried to do it before but there is an issue.
This code uses "sentEvent" which is an implementation in "AbstractAction". It is available in a private API in DatalakeDatabaseBackupActions.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+        return new SdxDatabaseBackupResponse(operationId);
+    }
+
+    public SdxDatabaseRestoreResponse triggerDatabaseRestore(SdxCluster sdxCluster, String backupId, String backupLocation) {
```
will update.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDrService.java:
```diff
@@ -0,0 +1,273 @@
+    @Inject
+    private SDxDatabaseDrStatusRepository sdxDatabaseDrStatusRepository;
+
+    public SdxDatabaseBackupResponse triggerDatabaseBackup(SdxCluster sdxCluster, String backupId, String backupLocation) {
```
will update.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:
```diff
@@ -104,6 +109,18 @@ public FlowIdentifier triggerSdxStopFlow(SdxCluster cluster) {
+        return notifyDatabaseDrEvent(selector, new DatalakeDatabaseBackupStartEvent(selector, sdxId, userId, backupId, backupLocation));
+    }
+
+    public String triggerDatalakeDatabaseRestoreFlow(Long sdxId, String backupId, String backupLocation) {
```
will update.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDatabaseDrService.java:
```diff
@@ -0,0 +1,309 @@
+     * @return true if status is successful, false otherwise.
+     */
+    private boolean isStackOrClusterDrStatusComplete(Status status) {
+        return (Status.BACKUP_FINISHED.equals(status) ||
+                Status.RESTORE_FINISHED.equals(status)) ? true : false;
```
```
return Status.BACKUP_FINISHED.equals(status) ||
                Status.RESTORE_FINISHED.equals(status);
```

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDatabaseDrService.java:
```diff
@@ -0,0 +1,309 @@
+     * @return true if status is failed, false otherwise.
+     */
+    private boolean isStackOrClusterStatusFailed(Status status) {
+        return (Status.BACKUP_FAILED.equals(status) ||
+                Status.RESTORE_FAILED.equals(status)) ? true : false;
```
```
return Status.BACKUP_FAILED.equals(status) ||
                Status.RESTORE_FAILED.equals(status);
```

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDatabaseDrService.java:
```diff
@@ -0,0 +1,309 @@
+        return AttemptResults.finishWith(stackV4Response);
+    }
+
+    /**
+     * Checks the status if the backup/restore operation is successful.
+     * @param status It could be stack/cluster status
+     * @return true if status is successful, false otherwise.
+     */
```
drop the comment, method name is self explaining

kkalvagadda1 replied:
will do

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDatabaseDrService.java:
```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    /**
+     * Checks the status if the backup/restore operation failed.
+     * @param status It could be stack/cluster status
+     * @return true if status is failed, false otherwise.
+     */
```
drop the comment

kkalvagadda1 replied:
will do

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/entity/SdxDatabaseDrStatus.java:
```diff
@@ -0,0 +1,92 @@
+
+@Entity
+@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"operationId"}))
+public class SdxDatabaseDrStatus {
```
As per our conversation, I will generalize the classes and move them to a new package,

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxDatabaseDrService.java:
```diff
@@ -0,0 +1,309 @@
+    @Inject
+    private SDxDatabaseDrStatusRepository sdxDatabaseDrStatusRepository;
+
+    public SdxDatabaseBackupResponse triggerDatabaseBackup(SdxCluster sdxCluster, String backupId, String backupLocation) {
+
+        return triggerDatalakeDatabaseBackupFlow(sdxCluster.getId(), backupId, backupLocation);
+    }
+
+    public SdxDatabaseRestoreResponse triggerDatabaseRestore(SdxCluster sdxCluster, String backupId, String backupLocation) {
+        return triggerDatalakeDatabaseRestoreFlow(sdxCluster.getId(), backupId, backupLocation);
+    }
```
will do

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/DatalakeDatabaseDrState.java:
```diff
@@ -0,0 +1,35 @@
+import com.sequenceiq.flow.core.RestartAction;
+import com.sequenceiq.flow.core.restart.DefaultRestartAction;
+
+public enum DatalakeDatabaseDrState implements FlowState {
```
will do.

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/event/DatalakeDatabaseBackupStartEvent.java:
```diff
@@ -0,0 +1,25 @@
+
+public class DatalakeDatabaseBackupStartEvent extends DatalakeDatabaseDrStartBaseEvent {
+
+    private String backupId;
+
+    private String backupLocation;
```
will fix

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/event/DatalakeDatabaseBackupSuccessEvent.java:
```diff
@@ -0,0 +1,17 @@
+import com.sequenceiq.datalake.flow.SdxEvent;
+
+public class DatalakeDatabaseBackupSuccessEvent extends SdxEvent {
+    private String operationId;
```
will fix

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/event/DatalakeDatabaseDrStartBaseEvent.java:
```diff
@@ -0,0 +1,19 @@
+import com.sequenceiq.datalake.flow.SdxEvent;
+
+public class DatalakeDatabaseDrStartBaseEvent extends SdxEvent  {
+    private SdxDatabaseDrStatus drStatus;
```
will fix

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/event/DatalakeDatabaseRestoreStartEvent.java:
```diff
@@ -0,0 +1,25 @@
+import com.sequenceiq.datalake.entity.SdxDatabaseDrOperationTypeEnum;
+
+public class DatalakeDatabaseRestoreStartEvent extends DatalakeDatabaseDrStartBaseEvent {
+    private String backupId;
+
+    private String backupLocation;
```
will fix

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/handler/DatalakeDatabaseBackupWaitHandler.java:
```diff
@@ -0,0 +1,69 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DatalakeDatabaseBackupWaitHandler.class);
+
+    private static final int SLEEP_TIME_IN_SEC = 20;
+
+    private static final int DURATION_IN_MINUTES = 90;
```
I did find a place where these values are configurable. Can you provide a reference?

lacikaaa replied:
look for `@Value` anywhere in the code

kkalvagadda1 commented on datalake/src/main/java/com/sequenceiq/datalake/flow/dr/handler/DatalakeDatabaseRestoreWaitHandler.java:
```diff
@@ -0,0 +1,69 @@
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+
+@Component
+public class DatalakeDatabaseRestoreWaitHandler extends ExceptionCatcherEventHandler<DatalakeDatabaseRestoreWaitRequest> {
```
They send different events on success and failure scenarios. I think it will be clean to have them separate.

kkalvagadda1 commented on core/src/main/java/com/sequenceiq/cloudbreak/job/StackStatusCheckerJob.java:
```diff
@@ -142,8 +142,10 @@ protected void executeInternal(JobExecutionContext context) throws JobExecutionE
                 Status.BACKUP_IN_PROGRESS,
                 Status.BACKUP_FAILED,
+                Status.BACKUP_FINISHED,
                 Status.RESTORE_IN_PROGRESS,
-                Status.RESTORE_FAILED
+                Status.RESTORE_FAILED,
+                Status.RESTORE_FINISHED
```
will move the finished states to synchable states

kkalvagadda1 commented:
> 3 thing:
> 
> * if you would like your flows to be retryable you should add them to `SdxFlowInformation#RESTARTABLE_FLOWS`
> * you should separate your flow, 2 separate package, 2 separate state/event enum
> * please squash your commits on the pr, you can always use amend and force push to update your branch. this way we can ensure your pr would be merged in single commit, and won't rely on the merger to be aware of the multiple commits

I addressed your comments here. 
Let me clarify a few things for you.
1. Backup is a periodic activity and we are planning to recommend customers to perform the backup once every few hours so that the metadata is not way behind when a restore is performed.
2. You are right, if there is metadata constantly updated, they could be out of sync. It is a known limitation and will be documented. There are some use cases where putting some restrictions would be possible. For example: performing backup as part of an upgrade. we can make sure that services like HMS/Ranger/Atlas are stopped before performing a backup.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/converter/SdxDatabaseDrStatusTypeEnumConverter.java:
```diff
@@ -0,0 +1,12 @@
+import com.sequenceiq.cloudbreak.converter.DefaultEnumConverter;
+import com.sequenceiq.datalake.entity.operation.SdxOperationStatusTypeEnum;
+
+public class SdxDatabaseDrStatusTypeEnumConverter extends DefaultEnumConverter<SdxOperationStatusTypeEnum> {
```
I think this class should be renamed also

kkalvagadda1 replied:
ok

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/operation/SdxOperationStatus.java:
```diff
@@ -0,0 +1,92 @@
+
+@Entity
+@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"operationId"}))
+public class SdxOperationStatus {
```
I think `SdxOperation` would fit better, there is too much status around this class

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/operation/SdxOperationStatusTypeEnum.java:
```diff
@@ -0,0 +1,19 @@
+package com.sequenceiq.datalake.entity.operation;
+
+public enum SdxOperationStatusTypeEnum {
```
rename to `SdxOperationStatus`

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/entity/operation/SdxOperationTypeEnum.java:
```diff
@@ -0,0 +1,17 @@
+package com.sequenceiq.datalake.entity.operation;
+
+public enum SdxOperationTypeEnum {
```
drop the `Enum` from the end

lacikaaa commented on core/src/main/java/com/sequenceiq/cloudbreak/job/StackStatusCheckerJob.java:
```diff
@@ -156,7 +156,9 @@ protected void executeInternal(JobExecutionContext context) throws JobExecutionE
                 Status.STOP_FAILED,
-                Status.AMBIGUOUS
+                Status.AMBIGUOUS,
+                Status.BACKUP_FINISHED,
+                Status.RESTORE_FINISHED
```
I don't know which pr will contain the final setting for this, but the 2 FAILED statuses should be here also.
cc @hreeve-cloudera

kkalvagadda1 replied:
will take care of it.

kkalvagadda1 commented:
> SdxClusterStatusCheckerJob

It is used to update the status of the datalake and also to notify the status change of datalake.
we don't want to change the status of datalake when the database backup/restore is performed. There will be another patch which will do this when the datalake backup/restore are performed. This will be triggered by datalake-dr service.
